fhirr4/ballerina/src/main/resources/fhirservice/fhir_preprocessor.bal

```diff
@@ -166,6 +181,293 @@ public isolated class FHIRPreprocessor {
         setFHIRContext(fhirCtx, httpCtx);
     }
 
+    # Process the FHIR Instance History interaction.
+    #
+    # + fhirResourceType - The FHIR resource type
+    # + id - The FHIR resource id
+    # + httpRequest - The HTTP request
+    # + httpCtx - The HTTP request context
+    # + return - Error if occurs
+    public isolated function processInstanceHistory (string fhirResourceType, string id, http:Request httpRequest, http:RequestContext httpCtx)
+                                                                                    returns r4:FHIRError? {
+        log:printDebug("Pre-processing FHIR interaction : instance history");
+        // Validate main HTTP headers
+        r4:FHIRRequestMimeHeaders clientHeaders = check validateClientRequestHeaders(httpRequest);
+
+        // Create interaction
+        readonly & FHIRInstanceHistoryInteraction historyInteraction = {id: id};
+
+        // Create FHIR request
+        r4:FHIRRequest fhirRequest = new (historyInteraction, fhirResourceType, (), {}, clientHeaders.acceptType);
+
+        // Populate JWT information in FHIR context
+        readonly & r4:FHIRSecurity fhirSecurity = check getFHIRSecurity(httpRequest);
+
+        // Handle SMART security
+        if fhirResourceType == PATIENT_RESOURCE {
+            r4:FHIRError? handleSmartSecurityResult = self.handleSmartSecurity(fhirSecurity, id);
+            if handleSmartSecurityResult is r4:FHIRError {
+                return handleSmartSecurityResult;
+            }
+        }
```

Comment:
Can we extract out to a common step before processing interactions?.seems same logic is everywhere?

Reply:
Smart security is handled only where patient id is involved. Also, we need the fhir security record to handle smart security and populating that record is done within the preprocessing. So having it in all the preprocessors would be suitable IMO. Otherwise we need could have it in the HTTP service builder which may need considerable if-else's.